mongo/collection.go

```diff
@@ -1670,6 +1669,32 @@ func makePinnedSelector(sess *session.Client, defaultSelector description.Server
 	}
 }
 
+// makePinnedOutputSelector makes a selector for a pinned session with a pinned server given that the underlying operation
+// is an aggregate with a $out or $merge stage. Will attempt to do server selection on the pinned server. If that fails, will
+// remove secondaries with wire version < 13 from the list of candidates, and will go through a list of default selectors.
+func makePinnedOutputSelector(sess *session.Client, defaultSelector description.ServerSelector) description.ServerSelectorFunc {
+	return func(t description.Topology, svrs []description.Server) ([]description.Server, error) {
+		if sess != nil && sess.PinnedServer != nil {
+			// If there is a pinned server, try to find it in the list of candidates.
+			for _, candidate := range svrs {
+				if candidate.Addr == sess.PinnedServer.Addr {
+					return []description.Server{candidate}, nil
+				}
+			}
+
+			return nil, nil
+		}
+
+		var cands []description.Server
+		for _, svr := range svrs {
+			if svr.Kind != description.RSSecondary || svr.WireVersion.Max >= 13 {
+				cands = append(cands, svr)
+			}
+		}
```

Comment:
`makePinnedOutputSelector` is a copy of `makePinnedSelector`, except it removes all secondaries with wire version < 13 from the list of candidates with the 5 lines above.

Let's say a user executes an aggregate with a `$out` stage and a `Secondary` read preference. Let's also say they execute this aggregate on a three-node replica set where each node supports a maximum wire version of 10. The `ServerSelectorFunc` returned by `makePinnedOutputSelector` will remove both secondaries from the candidates list because their wire versions are too low. `defaultSelector.SelectServer(t, cands)` below will honor the `Secondary` read preference, and will not return the primary as a candidate server. So, server selection will fail.

It is unclear to me whether the spec defines this as expected behavior. I've just opened a thread on the spec PR.